kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/KeyValueHeaderRecordWriterProvider.java

```diff
@@ -75,64 +75,86 @@ public RecordWriter getRecordWriter(S3SinkConnectorConfig conf, String filename)
         ? filename.substring(0, filename.length() - valueProvider.getExtension().length())
         : filename;
 
-    RecordWriter valueWriter = valueProvider.getRecordWriter(conf, strippedFilename);
-    RecordWriter keyWriter =
-        keyProvider == null ? null : keyProvider.getRecordWriter(conf, strippedFilename);
-    RecordWriter headerWriter =
-        headerProvider == null ? null : headerProvider.getRecordWriter(conf, strippedFilename);
+    Optional<RecordWriter> valueWriter = filename.contains(conf.getTombstoneEncodedPartition())
+        ? Optional.empty() : Optional.of(valueProvider.getRecordWriter(conf, strippedFilename));
+    Optional<RecordWriter> keyWriter = Optional.ofNullable(keyProvider)
+            .map(keyProvider -> keyProvider.getRecordWriter(conf, strippedFilename));
+    Optional<RecordWriter> headerWriter = Optional.ofNullable(headerProvider)
+            .map(headerProvider -> headerProvider.getRecordWriter(conf, strippedFilename));
 
     return new RecordWriter() {
       @Override
       public void write(SinkRecord sinkRecord) {
+        if (conf.isTombstoneWriteEnabled() && !keyWriter.isPresent()) {
+          throw new ConnectException(
+              "Key Writer must be configured when writing tombstone records is enabled.");
+        }
+
         // The two data exceptions below must be caught before writing the value
         // to avoid misaligned K/V/H files.
 
         // keyWriter != null means writing keys is turned on
-        if (keyWriter != null && sinkRecord.key() == null) {
+        if (keyWriter.isPresent() && sinkRecord.key() == null) {
           throw new DataException(
               String.format("Key cannot be null for SinkRecord: %s",
                   sinkRecordToLoggableString(sinkRecord))
           );
         }
 
         // headerWriter != null means writing headers is turned on
-        if (headerWriter != null
+        if (headerWriter.isPresent()
             && (sinkRecord.headers() == null || sinkRecord.headers().isEmpty())) {
           throw new DataException(
               String.format("Headers cannot be null for SinkRecord: %s",
                   sinkRecordToLoggableString(sinkRecord))
           );
         }
 
-        valueWriter.write(sinkRecord); // null check happens in sink task
-        if (keyWriter != null) {
-          keyWriter.write(sinkRecord);
+        if (sinkRecord.value() == null) {
+          validateTombstoneWriteEnabled(sinkRecord);
+        } else {
```

Comment:
what if it is set to ignore?? even in that case -- null value will get written

Reply:
Ignore will skip the records early on in the sink tasks put() method. Such records wouldn't arrive here if they are configured to be ignored.